SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp

```diff
@@ -0,0 +1,278 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Interop/Connection/SpatialEventTracer.h"
+
+#include <inttypes.h>
+
+#include "Interop/Connection/SpatialTraceEventBuilder.h"
+#include "SpatialGDKSettings.h"
+#include <WorkerSDK/improbable/c_io.h>
+#include <WorkerSDK/improbable/c_trace.h>
+
+DEFINE_LOG_CATEGORY(LogSpatialEventTracer);
+
+using namespace SpatialGDK;
+using namespace worker::c;
+
+void SpatialEventTracer::TraceCallback(void* UserData, const Trace_Item* Item)
+{
+	SpatialEventTracer* EventTracer = static_cast<SpatialEventTracer*>(UserData);
+
+	Io_Stream* Stream = EventTracer->Stream.Get();
+	if (!ensure(Stream != nullptr))
+	{
+		return;
+	}
+
+	uint32_t ItemSize = Trace_GetSerializedItemSize(Item);
+	if (EventTracer->BytesWrittenToStream + ItemSize <= EventTracer->MaxFileSize)
+	{
+		EventTracer->BytesWrittenToStream += ItemSize;
+		int Code = Trace_SerializeItemToStream(Stream, Item, ItemSize);
+		if (Code != 1)
+		{
+			UE_LOG(LogSpatialEventTracer, Error, TEXT("Failed to serialize to with error code %d (%s"), Code, Trace_GetLastError());
+		}
+	}
+	else
+	{
+		EventTracer->BytesWrittenToStream = EventTracer->MaxFileSize;
+	}
+}
+
+SpatialScopedActiveSpanId::SpatialScopedActiveSpanId(SpatialEventTracer* InEventTracer, const TOptional<Trace_SpanId>& InCurrentSpanId)
+	: CurrentSpanId(InCurrentSpanId)
+	, EventTracer(InEventTracer->GetWorkerEventTracer())
+{
+	if (InCurrentSpanId.IsSet())
+	{
+		Trace_EventTracer_SetActiveSpanId(EventTracer, CurrentSpanId.GetValue());
+	}
+}
+
+SpatialScopedActiveSpanId::~SpatialScopedActiveSpanId()
+{
+	if (CurrentSpanId.IsSet())
+	{
+		Trace_EventTracer_ClearActiveSpanId(EventTracer);
+	}
+}
+
+SpatialEventTracer::SpatialEventTracer(const FString& WorkerId)
+{
+	if (const USpatialGDKSettings* Settings = GetDefault<USpatialGDKSettings>())
+	{
+		if (Settings->bEventTracingEnabled)
+		{
+			MaxFileSize = Settings->MaxEventTracingFileSizeBytes;
+			Enable(WorkerId);
+		}
+	}
+}
+
+SpatialEventTracer::~SpatialEventTracer()
+{
+	if (IsEnabled())
+	{
+		Disable();
+	}
+}
+
+TOptional<Trace_SpanId> SpatialEventTracer::TraceEvent(FSpatialTraceEvent SpatialTraceEvent)
+{
+	return TraceEvent(SpatialTraceEvent, nullptr, 0);
+}
+
+TOptional<Trace_SpanId> SpatialEventTracer::TraceEvent(FSpatialTraceEvent SpatialTraceEvent, const worker::c::Trace_SpanId Causes)
+{
+	return TraceEvent(SpatialTraceEvent, &Causes, 1);
+}
+
+TOptional<Trace_SpanId> SpatialEventTracer::TraceEvent(FSpatialTraceEvent SpatialTraceEvent, const TArray<worker::c::Trace_SpanId>& Causes)
+{
+	return TraceEvent(SpatialTraceEvent, Causes.GetData(), Causes.Num());
+}
+
+TOptional<Trace_SpanId> SpatialEventTracer::TraceEvent(FSpatialTraceEvent SpatialTraceEvent, const worker::c::Trace_SpanId* Causes,
+													   int32 NumCauses)
+{
+	if (!IsEnabled())
+	{
+		return {};
+	}
+
+	Trace_SpanId CurrentSpanId;
+	if (NumCauses > 0)
```

Comment:
Is it valid for this to be 0?

Reply:
Yes, it would be the starting event trace in the chain.